mirgecom/initializers.py

```diff
@@ -840,3 +841,149 @@ def __call__(self, x_vec, eos, *, t=0.0):
 
         return join_conserved(dim=self._dim, mass=mass, energy=energy,
                               momentum=mom, species_mass=specmass)
+
+
+class Discontinuity:
+    r"""Solution initializer for flow with a discontinuity.
+
+    This initializer creates a physics-consistent mixture solution
+    given an initial thermal state (pressure, temperature) and a
+    mixture-compatible EOS.
+
+    The solution varies across a planar interface defined by a tanh fucntion
+    located at x=xloc for density, velocity, and pressure
+
+    .. automethod:: __init__
+    .. automethod:: __call__
+    """
+
+    def __init__(
+            self, *, dim=3, xdir=0, x0=0, nspecies=0,
+            tl=300.0, tr=600.0,
+            pl=1.e5, pr=2.e5,
+            ul=None, ur=None,
+            yl=None, yr=None,
+            uc=None, sigma=0.5
```

Comment:
I _like_ the brevity, but experience indicates we should be more descriptive with parameter names - at least in the actual user-facing interface.  Internally, I think it is less important because only those who (think they) know what they are doing will monkey with it internally.

Reply:
I agree. Look at all that nice empty space you have there to fill up. 🙂

Reply:
I agree, by default I use left and right, this is obvious when the discontinuity is in the x-direction, but less so in y and z? I'm ok with it if the group is though. I like some directionality over things like t1 and t2 for example. in this case it's not clear which side of the interface is 1 or 2. I could also be talked into plus and minus.

How verbose should/can we be?
temperature_<left/right>, pressure<left/right>, velocity<left/right>, species_mass_<left/right>

I think preserving the shortened names helps readability in the call routine.

Reply:
What do we think about something like this?
Tweak to interface: anderson2981/mirgecom#1
Corresponding call site: anderson2981/CEESD-Y1_Nozzle#1

Reply:
For me as a user, this seems less intuitive. What are we gaining by using tuples? I could see the advantage if it was a common data structure in the code, but this is a weird mix of conserved and dependent variables. If we could agree on a uniform API for all the initializers, then that would be something to consider further I think. We would need to think hard about how the data should be passed in, the needs vary between the different initializers.

Reply:
> For me as a user, this seems less intuitive. What are we gaining by using tuples? I could see the advantage if it was a common data structure in the code, ...

It is a little bit messy having two non-descriptive variables for each left/right thing.  The tuples make the call smaller, and better associate the things that are coming in pairs.  I do agree that the changed interface is harder to "deal with" from both the user and init-developer perspective. 

For this particular init, I think it is fine if the variable names are just changed to be descriptive. In general, it would be nice if we had a natural programming dialect for things that come in pairs; a left and right thing, or a plus and minus thing.  

> ... but this is a weird mix of conserved and dependent variables. If we could agree on a uniform API for all the initializers, then that would be something to consider further I think. We would need to think hard about how the data should be passed in, the needs vary between the different initializers.

I am not sure it is entirely relevant whether the data are conserved and or dependent variables, or a mix of them.  The init routines need to support arbitrary input data, as their data needs are sharply different. That said, if we had a nice way to pass in arbitrary parameters to a routine (like **kwargs) or so, it'd be a snap to unify the init API.  

We should think about what init API unification would really buy us, however.  One would not be able to pass just any parameters willy-nilly and expect it to work, the user will still need to read the init doc (or use an example) to know the data requirements for the init.

Reply:
I think the goal is that the API should allow the user to look at an existing driver and tell exactly what the input parameters do. The descriptive variables names accomplish that I think. I'll make the change and push something.